### PR TITLE
feat: abstract the endpoints-related logic

### PR DESCRIPTION
**Pull request by tokers**
In this Pull Request, we abstracted all the enpoints-related logic to `kube.Endpoint`, which hides its real entity. These changes are used to implement the EndpointSlice controller.

Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues

___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

**Comment by gxthrj** [comment]
Any issue related？

**Comment by tokers** [comment]
@gxthrj See #201 .

**Comment by gxthrj** [comment]
Please fix lint errors.

**Comment by tokers** [comment]
> Please fix lint errors.

Fixed.

**Review by tao12345666333 — APPROVED** [review]

File: pkg/config/config.go [hunk 1]
```diff
@@ -67,13 +67,14 @@ type Config struct {
...
-	IngressClass       string             `json:"ingress_class" yaml:"ingress_class"`
-	IngressVersion     string             `json:"ingress_version" yaml:"ingress_version"`
-	ApisixRouteVersion string             `json:"apisix_route_version" yaml:"apisix_route_version"`
+	Kubeconfig          string             `json:"kubeconfig" yaml:"kubeconfig"`
+	ResyncInterval      types.TimeDuration `json:"resync_interval" yaml:"resync_interval"`
+	AppNamespaces       []string           `json:"app_namespaces" yaml:"app_namespaces"`
+	ElectionID          string             `json:"election_id" yaml:"election_id"`
+	IngressClass        string             `json:"ingress_class" yaml:"ingress_class"`
+	IngressVersion      string             `json:"ingress_version" yaml:"ingress_version"`
+	WatchEndpointSlices bool               `json:"watch_endpoint_slices" yaml:"watch_endpoint_slices"`
```
Comment (gxthrj) [inline pkg/config/config.go:76]: Should add the configuration for (k8s.io/api/discovery/v1)LabelManagedBy ? or reuse the field `IngressClass`?

File: go.mod [hunk 2]
```diff
@@ -19,8 +21,8 @@ require (
 	go.uber.org/zap v1.13.0
 	golang.org/x/net v0.0.0-20210224082022-3d97a244fca7
 	gopkg.in/yaml.v2 v2.4.0
-	k8s.io/api v0.20.2
-	k8s.io/apimachinery v0.20.2
-	k8s.io/client-go v0.20.2
+	k8s.io/api v0.21.1
+	k8s.io/apimachinery v0.21.1
+	k8s.io/client-go v0.21.1
```
Comment (gxthrj) [inline go.mod:26]: Once we upgrade the version of  `client-go`,  Please make sure what is the minimum k8s version we  can support, and modify the doc.

**Review by gxthrj — COMMENTED** [review]
Great abstraction ~ with some small suggestions.

**Comment by gxthrj** [comment]
Will the test cases of endpointslice be added in the next PR?

**Comment by tokers** [comment]
> Will the test cases of endpointslice be added in the next PR?

I have changed the e2e framework to just use endpointslice instead of endpoints.

File: go.mod (hunk 2 above)
Comment (tokers) [reply go.mod:26]: OK.

File: pkg/config/config.go (hunk 1 above)
Comment (tokers) [reply pkg/config/config.go:76]: We only care `EndpointSlices` managed by `endpointslice-controller.k8s.io`.

```
endpointslice.kubernetes.io/managed-by: endpointslice-controller.k8s.io
```

**Comment by gxthrj** [comment]
> > Will the test cases of endpointslice be added in the next PR?
> 
> I have changed the e2e framework to just use endpointslice instead of endpoints.

I can only see the test for endpoints, such as [here](https://github.com/apache/apisix-ingress-controller/pull/574/files#diff-ac5b3974b052ded7c7a28e3658d27bcddf834d2a93168c0f1d25d6491f37cfc3R251)

File: go.mod (hunk 2 above)
Comment (tokers) [reply go.mod:26]: As per the compatibility matrix in https://github.com/kubernetes/client-go/tree/v0.21.1#compatibility-client-go---kubernetes-clusters, we should change the minimal supported k8s version to `1.15`.

**Review by gxthrj — APPROVED** [review]

**Pull request merged**